packages/api/src/external/carequality/document/process-document-query-results.ts

```diff
@@ -0,0 +1,167 @@
+import { MedicalDataSource } from "../../../external";
+import { appendDocQueryProgressWithSource } from "../../hie/append-doc-query-progress-with-source";
+
+const region = Config.getAWSRegion();
+const s3Utils = new S3Utils(region);
+const s3BucketName = Config.getMedicalDocumentsBucketName();
+
+export async function processDocumentQueryResults({
+  requestId,
+  patientId,
+  cxId,
+  documentQueryResults,
+}: {
+  requestId: string;
+  patientId: string;
+  cxId: string;
+  documentQueryResults: DocumentQueryResult[];
+}): Promise<void> {
+  const { log } = out(`CQ query docs - requestId ${requestId}, M patient ${patientId}`);
+
+  const docRefs = combineDocRefs(documentQueryResults);
+
+  try {
+    const docsToDownload = await getNonExistentDocRefs(docRefs, patientId, cxId);
+
+    const convertibleDocCount = docsToDownload.filter(doc =>
+      isConvertible(doc.contentType || undefined)
+    ).length;
+
+    log(`I have ${docsToDownload.length} docs to download (${convertibleDocCount} convertible)`);
+
+    await appendDocQueryProgressWithSource({
+      patient: { id: patientId, cxId: cxId },
+      downloadProgress: {
+        status: "processing",
+        total: docsToDownload.length,
+      },
+      convertProgress: {
+        status: "processing",
+        total: convertibleDocCount,
+      },
+      requestId,
+      source: MedicalDataSource.CAREQUALITY,
+    });
+
+    // TODO - INTRODUCED WHEN IMPLEMENTING CQ DOC RETRIEVAL
+    // NOTE MAKE SURE TO ADD DOC REFS TO TABLE WITH REQUESTID
+    // downloadDocs(docsToDownload, patientId, cxId, requestId);
+  } catch (error) {
+    const msg = `Failed to process documents in Carequality.`;
+    console.log(`${msg}. Error: ${errorToString(error)}`);
+
+    await appendDocQueryProgressWithSource({
+      patient: { id: patientId, cxId: cxId },
+      downloadProgress: { status: "failed" },
+      requestId,
+      source: MedicalDataSource.CAREQUALITY,
+    });
+
+    capture.message(msg, {
+      extra: {
+        context: `cq.processingDocuments`,
+        error,
+        patientId: patientId,
+        requestId,
+        cxId,
+      },
+      level: "error",
+    });
+    throw error;
+  }
+}
+
+function combineDocRefs(documentQueryResults: DocumentQueryResult[]): DocumentReference[] {
+  return documentQueryResults.reduce((acc: DocumentReference[], curr) => {
+    const documentReferences = curr.data.documentReference ?? [];
+
+    return [...acc, ...documentReferences];
+  }, []);
+}
+
+const getNonExistentDocRefs = async (
+  documents: DocumentReference[],
+  patientId: string,
+  cxId: string
+): Promise<DocumentReference[]> => {
+  const [{ existingDocRefs, nonExistingDocRefs }, FHIRDocRefs] = await Promise.all([
+    filterOutExistingDocRefsS3(documents, patientId, cxId),
+    getDocRefsFromFHIR(cxId, patientId),
+  ]);
+
+  const foundOnStorageButNotOnFHIR = existingDocRefs.filter(
+    f => !FHIRDocRefs.find(d => d.id === f.docUniqueId)
```

Comment:
this is O(n^2) in worst case since:

- Array.prototype.filter is O(n) because it goes through each element in existingDocRefs, 
- and Inside the filter callback, Array.prototype.find is called, which in the worst case will iterate over all elements in FHIRDocRefs for each element in existingDocRefs, for a total o O(n^2)

If you did something like this:

```
const fhirDocRefIds = new Set(FHIRDocRefs.map(d => d.id));
const foundOnStorageButNotOnFHIR = existingDocRefs.filter(
  f => !fhirDocRefIds.has(f.docUniqueId)
);
```
The runtime would be O(n), since creating the set is O(n), and filtering is O(n), but  each call to fhirDocRefIds.has is O(1)

Reply:
also patients can have thousands of documents right? so like O(n^2) vs O(n) like actually matters here I think